src/server/handle.rs

```diff
@@ -0,0 +1,565 @@
+        future.map(|v| {
+                let mut resp = RawDeleteResponse::new();
+                if let Some(err) = extract_region_error(&v) {
+                    resp.set_region_error(err);
+                } else if let Err(e) = v {
+                    resp.set_error(format!("{}", e));
+                }
+                resp
+            })
+            .boxed()
+    }
+
+    fn Coprocessor(&self, p: Request) -> GrpcFutureSend<Response> {
+        RECV_MSG_COUNTER.with_label_values(&["coprocessor"]).inc();
+
+        let (cb, future) = make_callback();
+        self.end_point_worker
+            .lock()
+            .unwrap()
+            .schedule(EndPointTask::Request(RequestTask::new(p, cb)))
+            .unwrap();
+        future.boxed()
+    }
+
+    fn Raft(&self, s: GrpcStreamSend<RaftMessage>) -> GrpcFutureSend<Done> {
+        let ch = self.ch.lock().unwrap().clone();
+        s.for_each(move |msg| {
+                ch.send_raft_msg(msg).map_err(|_| GrpcError::Other("send raft msg fail"))
+            })
+            .and_then(|_| future::ok::<_, GrpcError>(Done::new()))
+            .boxed()
+    }
+
+    fn Snapshot(&self, s: GrpcStreamSend<SnapshotChunk>) -> GrpcFutureSend<Done> {
+        let token = Token(self.token.fetch_add(1, Ordering::SeqCst));
+        let sched = self.snap_scheduler.lock().unwrap().clone();
+        let sched2 = sched.clone();
+        s.for_each(move |mut chunk| {
+                if chunk.has_message() {
+                    sched.schedule(SnapTask::Register(token, chunk.take_message()))
+                        .map_err(|_| GrpcError::Other("schedule snap_task fail"))
+                } else if !chunk.get_data().is_empty() {
+                    // TODO: Remove PipeBuffer or take good use of it.
+                    let mut b = PipeBuffer::new(chunk.get_data().len());
+                    b.write_all(chunk.get_data()).unwrap();
+                    sched.schedule(SnapTask::Write(token, b))
+                        .map_err(|_| GrpcError::Other("schedule snap_task fail"))
+                } else {
+                    Err(GrpcError::Other("empty chunk"))
+                }
+            })
+            .then(move |res| {
+                let res = match res {
+                    Ok(_) => sched2.schedule(SnapTask::Close(token)),
+                    Err(e) => {
+                        error!("receive snapshot err: {}", e);
+                        sched2.schedule(SnapTask::Discard(token))
+                    }
+                };
+                future::result(res.map_err(|_| GrpcError::Other("schedule snap_task fail")))
+            })
+            .and_then(|_| future::ok::<_, GrpcError>(Done::new()))
+            .boxed()
+    }
+}
+
+use storage::txn::Error as TxnError;
+use storage::mvcc::Error as MvccError;
+use storage::engine::Error as EngineError;
+
+fn extract_region_error<T>(res: &storage::Result<T>) -> Option<RegionError> {
```

Comment:
These funcs are copied from https://github.com/pingcap/tikv/blob/master/src/server/kv.rs , there are some old tests, I'll take them back in later PR.